repository/src/lib.rs

```diff
@@ -156,13 +157,196 @@ impl<T: RawRepository> DistributedRepository<T> {
+                        .to_hash256()
+                        .to_string()
+                        .truncate(COMMIT_TITLE_HASH_DIGITS)
+                );
+                self.raw
+                    .create_branch(branch_name, branch_tip_commit_hash)
+                    .await?;
+            } else if let Commit::Block(block_header) = branch_tip_commit {
+                // Verify finalization proof
+                let fp_commit_hash = self.raw.locate_branch(FP_BRANCH_NAME.into()).await?;
+                let fp_semantic_commit = self.raw.read_semantic_commit(fp_commit_hash).await?;
+                let finalization_proof: FinalizationProof =
+                    serde_json::from_str(&fp_semantic_commit.body).unwrap();
+                let result = verify::verify_finalization_proof(block_header, &finalization_proof);
+                // If the finalization proof is right, we push it to a vector to check heights
+                // If we can't find right finalization proof, we create a new b# branch
+                match result {
+                    Ok(()) => {
+                        block_headers.push(block_header.clone());
+                        finalized_branches.push(branch);
+                    }
+                    Err(_) => {
+                        let branch_tip_commit_hash = self.raw.locate_branch(branch).await?;
+                        let branch_name = format!(
+                            "b-{:?}",
+                            branch_tip_commit
+                                .to_hash256()
+                                .to_string()
+                                .truncate(COMMIT_TITLE_HASH_DIGITS)
+                        );
+                        self.raw
+                            .create_branch(branch_name, branch_tip_commit_hash)
+                            .await?
+                    }
+                }
+            } else {
+                // Else, fast-forward branches
+                self.raw.move_branch(branch, branch_tip_commit_hash).await?;
+            }
+        }
+        // Check the finalized branch's tip block commit height and leave only the highest finalized one
+        let block_heights: Vec<u64> = block_headers
+            .iter()
+            .map(|blockheader| blockheader.height)
+            .collect();
+        let highest_block_height = block_heights.iter().max().unwrap();
+        let mut same_height_finalized_block_count = 0;
+        let mut survived_finalized_branch_index = None;
+        for (index, block_height) in block_heights.iter().enumerate() {
+            if highest_block_height == block_height {
+                same_height_finalized_block_count += 1;
+                survived_finalized_branch_index = Some(index);
+            } else {
+                let to_be_deleted_branch_name = finalized_branches.get(index).unwrap().clone();
+                self.raw.delete_branch(to_be_deleted_branch_name).await?;
+            }
+        }
```

Comment:
I thinks this entire process is not intuitive.
Instead of saving all the finalized branch and performing a complicated comparison at the end,
how about just checking the merge base and find whether it is fork, whenever you hit the In the line 290?
(and also moving the `finalized` branch every time.

Reply:
I changed the whole process. fixed.